### ISSUE TITLE
Tensor.masked_fill applying mask incorrectly?

### ISSUE THREAD

**Issue by dustincoding** [dotnet/TorchSharp#1227]
Good evening,

I'm creating a tril mask (lower triangular mask) for my self attention model and it seems to create the tril correctly. However, when trying to apply the mask fill over my tensor, it applies the transposed matrix that I would expect. I'm wondering if this is by design, if I am using the masked_fill function incorrectly, or if this is a bug. Here are some pictures of the example:

1) My weights tensor during the q * transpose of k tensor:
![image](https://github.com/dotnet/TorchSharp/assets/107263472/e3c4cab1-be30-410c-ab28-b54cbc7f0cfb)

2) My mask tensor made from mask = torch.tril(torch.ones(blockSize, blockSize, device: device)).@bool(); This appear correctly as I would expect the mask to be viewed.
![image](https://github.com/dotnet/TorchSharp/assets/107263472/1fbe63bb-6c07-416a-817a-8626052c9a3a)

3) Applying the mask to the weights tensor does logic I would not expect. weights = weights.masked_fill(mask, float.NegativeInfinity);
![image](https://github.com/dotnet/TorchSharp/assets/107263472/2d35f9bf-d00b-418c-ac60-c27b2aacc855)

I would expect that the upper triangular portion of the tensor display would contain the -infinity values and that the lower triangular portion would contain the self attention weights. If this is a bug in my code (which is probably is), then I apologize ahead of time. But I think that there is something more here that could potentially be a bug in the way the mask is applied versus the way it is applied in PyTorch. Please let me know if there are any questions I can help with.

`public AttentionHead(int n_embd, int headSize, double dropoutVal, int blockSize, Device device, int id) : base ($"AttentionHead{id}")
{
    key = Linear(n_embd, headSize, false, device);
    query = Linear(n_embd, headSize, false, device);
    value = Linear(n_embd, headSize, false, device);

    //mask will modify the sublayer to prevent positions from attending to subsequent positions (we only want the context of previous positions)
    // combined with output embeddings offset by one position ensures the predictions for position i can only depend on known outputs at positions less than i
    mask = torch.tril(torch.ones(blockSize, blockSize, device: device)).@bool();
    this.register_buffer("tril", mask);
    Debug.WriteLine($"{mask.str}");
    dropout = Dropout(dropoutVal);

    RegisterComponents();
}

public override Tensor forward(Tensor x)
{
    var k = key.forward(x);
    Debug.WriteLine(k.str());
    var q = query.forward(x);
    Debug.WriteLine(q.str());

    // perform the dot product of Q and transposed K (MAYBE LOOK AT THIS USING WRONG DIMS FOR TRANSPOSE 1 based dims as opposed to 0 based?), scaling by sqrt(dksize) counteracts the magnitude of dot products on large values of dk
    var weights = torch.matmul(q, k.transpose(-2, -1)) / Math.Sqrt(k.size(-1));
    Debug.WriteLine(weights.str());
    Debug.WriteLine(mask.str());
    // look into the mask potentially being null (setting to -inf turns off all illegal connections to prevent leftward information flow and preserve auto-regressive properties)
    weights = weights.masked_fill(mask, float.NegativeInfinity);
    Debug.WriteLine(weights.str());

    //coverts output to predicted next-token probabilities (weights)
    weights = torch.nn.functional.softmax(weights, dim: -1);
    Debug.WriteLine(weights.str());
    weights = dropout.forward(weights);
    Debug.WriteLine(weights.str());

    var v = value.forward(x);
    Debug.WriteLine(v.str());
    var output = torch.matmul(weights, v);
    Debug.WriteLine(output.str());
    return output;
}

private void RegisterComponents()
{
    this.register_module("key", key);
    this.register_module("query", query);
    this.register_module("value", value);
    this.register_module("dropout", dropout);
}`

**Comment by dustincoding** [comment]
Okay hang on, I realized this is a bug in my code where I need to define what it means grab the values in the TxT matrix that are == 0.
